src/login.ts

```diff
@@ -0,0 +1,12 @@
+import * as core from '@actions/core';
+import * as pulumiCli from './libs/pulumi-cli';
+
+export const login = async (cloudUrl: string, accessToken: string) => {  
+  if (cloudUrl) {
+    core.debug(`Logging into ${cloudUrl}`);
+    await pulumiCli.run('login', cloudUrl);
+  } else if (accessToken == '') {
+    core.debug(`Logging into Pulumi`);
+    await pulumiCli.run('login');
+  }
```

Comment:
If `cloudUrl` is not set and `accessToken` is non-empty, this function falls through and does nothing.

Reply:
This code will be more maintainable and easier to reason about if we

1. keep the special casing to a minimum
2. let the login command do the error handling.

To that end, I would recommend changing the `else if` block to an `else` and adding `--non-interactive` to all the login commands.  That way, we can guarantee that `pulumi login` will be called and that if it doesn't have the information needed to complete the login it will abort immediately (presumably with a useful error message) instead of waiting for input that Actions can't provide.

Reply:
Looking into it, [you were right previously](https://github.com/pulumi/actions/pull/717/files#r961207899). There are scenarios where the Pulumi CLI does not need to login before performing operations. 

For example, as long as a stack has been selected, this `Pulumi.yaml` file works without logging in:

```yml
name: test-tracing
runtime: nodejs
description: A minimal TypeScript Pulumi program
backend:
  url: file://./data
```
...because the backend is a local file.

As a result, I think we should skip the login step if `PULUMI_ACCESS_TOKEN` and `cloud-url` are both empty.

Reply:
> If cloudUrl is not set and accessToken is non-empty, this function falls through and does nothing.

You're right, I'm missing a `!` here. We want to login when the token is set, I believe.